src/Lecture4.hs

```diff
@@ -135,7 +140,33 @@ errors. We will simply return an optional result here.
 -}
 
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow = readRow . splitCSVLine
+   where readRow :: [String] -> Maybe Row
+         readRow cols
+            | length cols /= 3 = Nothing
+            | otherwise = Row <$> readProduct c1 <*> readTradeType c2 <*> readInt c3
+               where c1 = head cols
+                     c2 = cols !! 1
+                     c3 = last cols
+         readProduct :: String -> Maybe String
+         readProduct p
+            | dropSpaces p == "" = Nothing
+            | otherwise = Just p
+         readTradeType :: String -> Maybe TradeType
+         readTradeType trade = case dropSpaces trade of
+            "Sell" -> Just Sell
+            "Buy"  -> Just Buy
+            _      -> Nothing
+         readInt :: String -> Maybe Int
+         readInt int = readMaybe int >>= positiveOrNothing
+         positiveOrNothing :: Int -> Maybe Int
+         positiveOrNothing x = if x < 0 then Nothing else Just x
+
+splitCSVLine :: String -> [String]
+splitCSVLine string = go string ""
+   where go :: String -> String -> [String]
+         go [] acc      = [acc]
+         go (x: xs) acc = if x == ',' then acc : go xs "" else go xs (acc ++ [x])
```

Comment:
Adding to the end of the list is slow 🐌 You don't have a space leak due to `splitCSVLine` implementation because its line is small. But it could be a good small performance optimization 🙂 

You can use standard functions `span` or `break` to split in a single traversal:

* https://hackage.haskell.org/package/base-4.16.0.0/docs/Prelude.html#v:span